server/src/main/java/org/elasticsearch/transport/InboundPipeline.java

```diff
@@ -92,99 +90,96 @@ public void handleBytes(TcpChannel channel, ReleasableBytesReference reference)
+            final int bytesDecoded = decode(channel, reference);
+            if (bytesDecoded != 0) {
+                reference = reference.releasableSlice(bytesDecoded);
             } else {
-                try {
-                    forwardFragments(channel, fragments);
-                } finally {
-                    for (Object fragment : fragments) {
-                        if (fragment instanceof ReleasableBytesReference) {
-                            ((ReleasableBytesReference) fragment).close();
-                        }
-                    }
-                    fragments.clear();
-                }
+                break;
             }
         }
+        // if handling the messages didn't cause the channel to get closed and we did not fully consume the buffer retain it
+        if (isClosed == false && reference.length() > 0) {
+            pending.add(reference.retain());
+        }
     }
 
-    private void forwardFragments(TcpChannel channel, ArrayList<Object> fragments) throws IOException {
-        for (Object fragment : fragments) {
-            if (fragment instanceof Header) {
-                assert aggregator.isAggregating() == false;
-                aggregator.headerReceived((Header) fragment);
-            } else if (fragment instanceof Compression.Scheme) {
-                assert aggregator.isAggregating();
-                aggregator.updateCompressionScheme((Compression.Scheme) fragment);
-            } else if (fragment == InboundDecoder.PING) {
-                assert aggregator.isAggregating() == false;
-                messageHandler.accept(channel, PING_MESSAGE);
-            } else if (fragment == InboundDecoder.END_CONTENT) {
-                assert aggregator.isAggregating();
-                try (InboundMessage aggregated = aggregator.finishAggregation()) {
-                    statsTracker.markMessageReceived();
-                    messageHandler.accept(channel, aggregated);
+    private int decode(TcpChannel channel, ReleasableBytesReference reference) throws IOException {
+        return decoder.decode(channel, reference, this::forwardFragment);
+    }
+
+    private void doHandleBytesWithPending(TcpChannel channel) throws IOException {
+        do {
+            final int bytesDecoded;
+            if (pending.size() == 1) {
+                bytesDecoded = decode(channel, pending.peekFirst());
+            } else {
+                try (ReleasableBytesReference toDecode = getPendingBytes()) {
+                    bytesDecoded = decode(channel, toDecode);
                 }
+            }
+            if (bytesDecoded != 0 && isClosed == false) {
+                releasePendingBytes(bytesDecoded);
             } else {
-                assert aggregator.isAggregating();
-                assert fragment instanceof ReleasableBytesReference;
-                aggregator.aggregate((ReleasableBytesReference) fragment);
+                return;
```

Comment:
If `isClosed` can we assert that `pending` is now empty?

Reply:
Yea that should be fine, will add